sdk/resource/resource.go

```diff
@@ -17,30 +17,53 @@
 package resource
 
 import (
-	"reflect"
+	"sort"
+	"strings"
 
 	"go.opentelemetry.io/otel/api/core"
 )
 
 // Resource describes an entity about which identifying information and metadata is exposed.
 type Resource struct {
-	labels map[core.Key]core.Value
+	str    string
+	sorted []core.KeyValue
+	keys   map[core.Key]struct{}
 }
 
 // New creates a resource from a set of attributes.
 // If there are duplicates keys then the first value of the key is preserved.
 func New(kvs ...core.KeyValue) *Resource {
-	res := &Resource{
-		labels: map[core.Key]core.Value{},
-	}
+	res := &Resource{keys: make(map[core.Key]struct{})}
 	for _, kv := range kvs {
-		if _, ok := res.labels[kv.Key]; !ok {
-			res.labels[kv.Key] = kv.Value
+		// First key wins.
+		if _, ok := res.keys[kv.Key]; !ok {
+			res.keys[kv.Key] = struct{}{}
+			res.sorted = append(res.sorted, kv)
 		}
 	}
+	sort.Slice(res.sorted, func(i, j int) bool {
+		return res.sorted[i].Key < res.sorted[j].Key
+	})
+	res.str = buildResourceString(res.sorted)
 	return res
 }
 
+// String implements the Stringer interface and provides a reproducibly
+// hashable representation of a Resource.
+func (r Resource) String() string {
+	return r.str
+}
+
+// Attributes returns a copy of attributes from the resource in a sorted order.
```

Comment:
Food for thought: Return `AttributeIterator` for cases someone needs to iterate over attributes without getting a copy of the slice. You could see what I did in `sdk/export/metric/metric.go` (the `LabelIterator` stuff).

Reply:
I didn't want to change the return type here given its use through the project so far, but I'm on board with potentially adding a `AttributeIterator` method. What are your thoughts on that?

Reply:
Looking into this more I like this idea, but I'm wondering if this kind of iterator pattern shouldn't be implemented alongside the `core.KeyValue` structure. I think I'm going to add this, but maybe in a subsequent PR we can unify the pattern.